### PR TITLE
[Test] fix flaky pg tests

### LINKED ISSUE #30573: [Test] gcs_placement_group_manager_test flaky in the master
### What happened + What you expected to happen

Screen Shot 2022-11-21 at 9 14 14 PM

After merging this PR, `test_placement_group` became flaky. We should investigate and fix it. 

cc @xwjiang2010 I will mark it as a release blocker until we find the root cause. 

cc @larrylian 

### Versions / Dependencies

master

### Reproduction script

n/a

### Issue Severity

_No response_

### PR DESCRIPTION
**Pull request by rkooo567**
Signed-off-by: SangBin Cho <rkooo567@gmail.com>

## Why are these changes needed?

From the gcs_pg_manager, every code is running on a single io service, meaning the class is not written in thread-safe way.

However (Little unfortunate), pg unit test starts a new thread to run io service (so there are 2 threads, main thread & io service thread). This causes the threading issue because if we access the same field from io service thread and main service thread, it causes the issue.

This PR temporarily fixes the issue by waiting until the io service is executed from the main thread. (alternatively, we can run all the APIs from the io service thread). I think fixing this holistically requires too much code change, and it is not worth it now. 

This PR also splits the pg tests into two tests that make more sense. But I also found the current code is pretty badly written, and I am 99% sure if is not working well under real node failures. I think we should rewrite some code & add more unit tests & chaos test. I will probably follow up in the beg of next year to do this. 

## Related issue number

Closes https://github.com/ray-project/ray/issues/30573#issue-1459099683

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/gcs/gcs_server/test/gcs_placement_group_manager_test.cc [hunk 1]
```diff
@@ -452,24 +512,20 @@ TEST_F(GcsPlacementGroupManagerTest, TestRescheduleWhenNodeDead) {
   ASSERT_EQ(placement_group->GetState(), rpc::PlacementGroupTableData::RESCHEDULING);
 
   // Test placement group rescheduling success.
+  mock_placement_group_scheduler_->placement_groups_.pop_back();
   OnPlacementGroupCreationSuccess(placement_group);
   ASSERT_EQ(placement_group->GetState(), rpc::PlacementGroupTableData::CREATED);
-  WaitForExpectedPgCount(1);
```
Comment (rkooo567) [inline src/ray/gcs/gcs_server/test/gcs_placement_group_manager_test.cc:457]: This part of test makes no sense and uses APIs incorrectly. I moved this part to another test

File: src/ray/gcs/gcs_server/test/gcs_placement_group_manager_test.cc [hunk 2]
```diff
@@ -166,6 +166,20 @@ class GcsPlacementGroupManagerTest : public ::testing::Test {
     EXPECT_TRUE(WaitForCondition(condition, 10 * 1000));
   }
 
+  void WaitUntilIoServiceDone() {
+    // In this test, io service is running in a different thread.
+    // That means it can have a thread safety issue, if the test thread
+    // touches private attributes while io service is accessing it.
+    // This method returns when there's no more task in the io service,
```
Comment (cadedaniel) [inline src/ray/gcs/gcs_server/test/gcs_placement_group_manager_test.cc:173]: What prevents new tasks from being added after this function runs?

Comment (cadedaniel) [reply src/ray/gcs/gcs_server/test/gcs_placement_group_manager_test.cc:173]: (fine if there's nothing atm, just trying to understand the nature of the fix)

Comment (rkooo567) [reply src/ray/gcs/gcs_server/test/gcs_placement_group_manager_test.cc:173]: Hmm I guess theoretically new task can be added if the API post another task within a task. But this case doesn't exist now

Comment (rkooo567) [reply src/ray/gcs/gcs_server/test/gcs_placement_group_manager_test.cc:173]: But this is the blocking call, and post ensures the insertion ordering, so normally it'd work

**Comment by rkooo567** [comment]
cc for review?

File: src/ray/gcs/gcs_server/test/gcs_placement_group_manager_test.cc (hunk 2 above)
Comment (cadedaniel) [reply src/ray/gcs/gcs_server/test/gcs_placement_group_manager_test.cc:173]: Awesome

**Review by cadedaniel — APPROVED** [review]

**Pull request merged**